modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replicator/IndexBuilderTxRwOperationTrackerTest.java

```diff
@@ -0,0 +1,171 @@
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.concurrent.CompletableFuture;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.RepeatedTest;
+import org.junit.jupiter.api.Test;
+
+/** For {@link IndexBuilderTxRwOperationTracker} testing. */
+public class IndexBuilderTxRwOperationTrackerTest {
+    /** To be used in a loop. {@link RepeatedTest} has a smaller failure rate due to recreating the tracker every time. */
+    private static final int REPEATS = 100;
+
+    private final IndexBuilderTxRwOperationTracker tracker = new IndexBuilderTxRwOperationTracker();
+
+    @AfterEach
+    void tearDown() {
+        tracker.close();
+    }
+
+    @Test
+    void testAwaitCompleteTxRwOperationsNoOperations() {
+        CompletableFuture<Void> awaitFuture0 = tracker.awaitCompleteTxRwOperations(0);
+        CompletableFuture<Void> awaitFuture1 = tracker.awaitCompleteTxRwOperations(1);
+        CompletableFuture<Void> awaitFuture2 = tracker.awaitCompleteTxRwOperations(1);
+
+        assertFalse(awaitFuture0.isDone());
+        assertFalse(awaitFuture1.isDone());
+        assertFalse(awaitFuture2.isDone());
+
+        tracker.updateMinAllowedCatalogVersionForStartOperation(0);
+
+        assertTrue(awaitFuture0.isDone());
+        assertFalse(awaitFuture1.isDone());
+        assertFalse(awaitFuture2.isDone());
+
+        tracker.updateMinAllowedCatalogVersionForStartOperation(1);
+
+        assertTrue(awaitFuture1.isDone());
+        assertTrue(awaitFuture2.isDone());
+    }
+
+    @Test
+    void testAwaitCompleteTxRwOperationsWithOperations() {
+        tracker.incrementOperationCount(0);
+        tracker.incrementOperationCount(1);
+
+        CompletableFuture<Void> awaitFuture0 = tracker.awaitCompleteTxRwOperations(0);
+        CompletableFuture<Void> awaitFuture1 = tracker.awaitCompleteTxRwOperations(1);
+
+        tracker.updateMinAllowedCatalogVersionForStartOperation(0);
+
+        assertTrue(awaitFuture0.isDone());
+        assertFalse(awaitFuture1.isDone());
+
+        tracker.decrementOperationCount(1);
+        assertFalse(awaitFuture1.isDone());
+
+        tracker.updateMinAllowedCatalogVersionForStartOperation(1);
+        assertFalse(awaitFuture1.isDone());
+
+        tracker.decrementOperationCount(0);
+        assertTrue(awaitFuture1.isDone());
+    }
+
+    @Test
+    void testRejectTxRwOperation() {
+        tracker.updateMinAllowedCatalogVersionForStartOperation(1);
+
+        assertThrows(RejectTxRwOperationException.class, () -> tracker.incrementOperationCount(0));
+
+        assertDoesNotThrow(() -> tracker.incrementOperationCount(1));
+        assertDoesNotThrow(() -> tracker.incrementOperationCount(2));
+    }
+
+    @Test
+    void testConcurrentUpdateMinAllowedAndAwaitCompleteOperations0() {
```

Comment:
Is zero at the end just an index of the test? If this is true, let's give the test a better name.

Reply:
Yes, this is just a test index, I don’t know which name is better to give, I suggest leaving it as it is or suggest a better name.

Reply:
`testMultipleConcurrentUpdateMinAllowedAndAwaitCompleteOperations`, for example, as this tests differs from its friends by doing multiple updates.

Reply:
This results in a rather long method name, which does not greatly improve understanding.
The test is short and clear, I suggest leaving it as it is for now, maybe a better naming option will appear later.

The current dialogue is like a matter of taste.

Reply:
If we have a few tests, they probably test different things, right? If it is true, then they should have different names (and the difference must be clear from the name, not just 0/1/2). If, on the other hand, they all test the same thing, then we don't need a few method, we just need one.